### PR TITLE
First attempt to move to path for states.

### LINKED ISSUE #65: H5PYDataset does not play well with DataStream for serialization
This discussion started at the [mailing list](https://groups.google.com/forum/#!topic/blocks-users/E01_gG7hPD4).

`ref_sources` is a class attribute in `H5PYDataset`, which causes it to not be serialized. However, if you use a `DataStream` in your code, a handle to the HDF5 file is saved as an instance attribute, which causes it to be serialized. However, most objects in h5py can't be serialized (more info [here](https://github.com/h5py/h5py/issues/531)), which causes loading a pickled `DataStream` to fail. I am not sure on what could be done, as saving the DataStream state is useful when the dataset is not an `H5PYDataset`. The handle to the HDF5 file does indeed hold a state while the file is loaded (which driver is being used, buffering settings, etc), but this information is not useful if reloading an HDF5 file from scratch.

### PR DESCRIPTION
**Pull request by Thrandis**
Here is a first draft of my code to fix #65. The role of open() seems a little weird, because even if the file has not be opened, we can still get data using get_file_handle(). Same for close(). What do you think @vdumoulin?

I also moved from dataset.close(handle) to dataset.close(), since we don't need to pass the handle anymore, is it okay?

### CODE REVIEW

File: fuel/datasets/hdf5.py [hunk 1]
```diff
@@ -119,7 +119,8 @@ class H5PYDataset(Dataset):
         indices are ordered.
 
     """
-    ref_counts = defaultdict(int)
+    _ref_counts = defaultdict(int)
+    _file_handles = defaultdict()
```
Comment (vdumoulin) [inline fuel/datasets/hdf5.py]: No need to make this a `defaultdict`. The reason why `_ref_counts` is a `defaultdict` is to simplify the logic in the defunct `_get_file_id` method, but for `_file_handles` we gain nothing.

File: fuel/datasets/hdf5.py [hunk 2]
```diff
@@ -300,20 +306,12 @@ def open(self):
...
     def _out_of_memory_open(self):
-        file_id = self._get_file_id()
-        state = h5py.File(name=file_id, mode="r", driver=self.driver)
-        self.ref_counts[state.id] += 1
-        return state
+        self.get_file_handle()
+        return self.path
 
-    def close(self, state):
+    def close(self):
```
Comment (vdumoulin) [inline fuel/datasets/hdf5.py]: This method is part of the interface defined in `Dataset` (see its definition [here](https://github.com/bartvm/fuel/blob/master/fuel/datasets/base.py#L130-L132)), and as such we should keep the `state` argument.

**Comment by vdumoulin** [comment]
You do have a point: passing the file path as the handle is pretty reduntant, and we could just as well pass `None`.

Here's an alternative:
* We put the code from `get_file_handle` back in `_out_of_memory_open` (except that it doesn't return anything) and we remove the `get_file_handle` method.
* The `open` method calls `self._out_of_memory_open()` if `self.load_in_memory` is `False` and returns `None` no matter what.
* We introduce the following property:
``` python
@property
def _file_handle(self):
    if not self.path in self._file_handles:
        raise IOError('no open handle for file {}'.format(self.path))
    return self._file_handles[self.path]
```
* File handles are accessed like so:
``` python
handle = self._file_handle
```
* We put the implementation of `release_file_handle` into `out_of_memory_close` and remove the `release_file_handle` method.

**Comment by vdumoulin** [comment]
It would also be good to add a unit test that fails currently but passes with your changes.

Try adding a test like
``` python
# ...
dataset = H5PYDataset(...)
stream = DataStream(dataset)
cPickle.dumps(stream)
```

**Comment by Thrandis** [comment]
@vdumoulin: The problem now is that when we instantiate a new H5PYDataset, the function split_dict() is called (through available_splits()), but the dataset is not opened yet.
Is it okay to call open() at this moment (or _out_of_memory_open()), and immediately close it afterwards?

**Comment by vdumoulin** [comment]
Yes, that would be acceptable: you're opening the file to read the metadata and then closing it.

**Comment by Thrandis** [comment]
I still have some tests that fails. It seems that I have trouble getting the data or when there is two different H5PYDataset on the same hdf5 file.

**Comment by vdumoulin** [comment]
Which tests, exactly?

File: fuel/datasets/hdf5.py [hunk 3]
```diff
@@ -275,12 +270,11 @@ def subsets(self):
                 if num_examples != subset.stop - subset.start:
                     raise ValueError("sources have different lengths")
             self._subsets = subsets
-            self._out_of_memory_close(handle)
         return self._subsets
 
     def load(self):
         if self.load_in_memory:
-            handle = self._out_of_memory_open()
```
Comment (vdumoulin) [inline fuel/datasets/hdf5.py]: In this specific case you should open and close: we're reading the file from disk and into memory. This is probably why your tests are failing.

**Comment by Thrandis** [comment]
test_hdf5.test_h5py_dataset_split
  -> ValueError: 'test' split is not provided by this dataset. Available splits are (u'train',).
test_hdf5.test_h5py_dataset_out_of_memory
 -> ValueError: 'test' split is not provided by this dataset. Available splits are (u'train',).
test_hdf5.test_h5py_flatten_in_memory
 -> AssertionError: Arrays are not equal
test_hdf5.test_h5py_flatten_out_of_memory
 -> AssertionError: Arrays are not equal

**Comment by vdumoulin** [comment]
Weird. Could you fix what I've outlined just to see if it solves the issue?

**Comment by Thrandis** [comment]
It don't solve the issuse ... :-(

**Comment by vdumoulin** [comment]
Calling `self._out_of_memory_open` and `self._out_of_memory_close` in `load` seems to solve at least part of the failures for me.

**Comment by Thrandis** [comment]
Mhhh strange. For me the 4 tests don't pass using `self._out_of_memory_open` and `self._out_of_memory_close`.

**Comment by vdumoulin** [comment]
You need to open and close in `load`.

I think I found why the tests fail: in `test_dataset_datastream_pickles`, you're not closing the data stream, which leaves a dangling file handle in the `_file_handles` dict. Since we're re-using `'tmp.hdf5'` everywhere in the tests, it causes some issues.

The solution is to make sure that `close` gets called no matter what, by putting it in the `finally` block. (By the way, it would be a good thing to do that for all tests in that file as well.)

**Comment by Thrandis** [comment]
Great! I'll change that!

**Comment by vdumoulin** [comment]
Looks good, thanks a lot!

**Comment by Thrandis** [comment]
@vdumoulin thanks for your help ;-)

**Pull request merged**